### PR TITLE
Some changes to "abstract".

### PR DESCRIPTION
**Pull request by chriseth**

### CODE REVIEW

**Review by ekpyron — APPROVED** [review]
So we *do* want to allow and require ``abstract`` for libraries? What was @axic's take on that? If we want to go with that, the PR is fine in any case. And I guess it makes sense this way.

**Review by ekpyron — CHANGES_REQUESTED** [review]
Has more tests with libraries that needs fixes now, apparently.

**Comment by axic** [comment]
Since this is only used for the case when someone wants to ship an incomplete library for build time (e.g. keeping code proprietary?), my opinion is to disallow it. It is easier to enable if there is a strong request for this feature, but would actually prefer to design the language better to accommodate this (perhaps using interfaces and a `using as library` statement?)

**Comment by chriseth** [comment]
Ok, I'll disallow it.

File: test/libsolidity/syntaxTests/controlFlow/storageReturn/unimplemented_library.sol [hunk 1]
```diff
@@ -1,4 +1,12 @@
...
 library L {
     function f() public returns(uint[] storage);
     function g() public returns(uint[] storage s);
 }
+abstract library T {
+    function f() public returns(uint[] storage);
+    function g() public returns(uint[] storage s);
+}
+// ----
+// TypeError: (32-145): Contract "L" should be marked as abstract.
```
Comment (ekpyron) [inline test/libsolidity/syntaxTests/controlFlow/storageReturn/unimplemented_library.sol]: Not the best error message...

Comment (ekpyron) [reply test/libsolidity/syntaxTests/controlFlow/storageReturn/unimplemented_library.sol]: Should be easy to change this to ``library functions must be implemented`` or ``Abstract libraries are not (yet) supported``? If that isn't easy, then it's probably not too bad, though...

Comment (chriseth) [reply test/libsolidity/syntaxTests/controlFlow/storageReturn/unimplemented_library.sol]: I know, that's what @axic wanted ;)

Comment (ekpyron) [reply test/libsolidity/syntaxTests/controlFlow/storageReturn/unimplemented_library.sol]: Yeah, but I merely meant that the error message probably shouldn't suggest to mark it "abstract", if we don't allow doing that :-).

**Comment by chriseth** [comment]
Changed it again.

**Comment by ekpyron** [comment]
Needs rebase.

**Comment by chriseth** [comment]
Rebased.

File: test/libsolidity/syntaxTests/nameAndTypeResolution/511_library_function_without_implementation_public.sol [hunk 2]
```diff
@@ -1,4 +1,7 @@
+// This used to work pre-0.6.0.
 library L {
     // This can be used as an "interface", hence it is allowed.
```
Comment (ekpyron) [inline test/libsolidity/syntaxTests/nameAndTypeResolution/511_library_function_without_implementation_public.sol:3]: This isn't exactly true anymore now :-).

**Review by ekpyron — COMMENTED** [review]
I'm still not *entirely* sure that it's a good idea to disallow "library interfaces".

**Review by ekpyron — APPROVED** [review]

**Pull request merged**